src/utils/webrtc/simplewebrtc/peer.js

```diff
@@ -372,22 +400,134 @@ Peer.prototype.end = function() {
 	this.pc.close()
 	this.handleStreamRemoved()
 	this.parent.off('localTrackReplaced', this.handleLocalTrackReplacedBound)
+	this.parent.off('localTrackEnabledChanged', this.handleLocalTrackEnabledChangedBound)
 }
 
 Peer.prototype.handleLocalTrackReplaced = function(newTrack, oldTrack, stream) {
+	this._pendingReplaceTracks.push({ newTrack, oldTrack, stream })
+
+	this._processPendingReplaceTracks()
+}
+
+/**
+ * Process pending replace track actions.
+ *
+ * All the pending replace track actions are executed from the oldest to the
+ * newest, waiting until the previous action was executed before executing the
+ * next one.
+ *
+ * The process may be stopped if the connection is lost, or if a track needs to
+ * be added rather than replaced, which requires a renegotiation. In both cases
+ * the process will start again once the connection is restablished.
+ *
+ * @returns {Promise} a Promise fulfilled when the processing ends; if it was
+ *          completed the resolved value is true, and if it was stopped before
+ *          finishing the resolved value is false.
+ */
+Peer.prototype._processPendingReplaceTracks = function() {
+	if (this._processPendingReplaceTracksPromise) {
+		return this._processPendingReplaceTracksPromise
+	}
+
+	this._processPendingReplaceTracksPromise = new Promise((resolve, reject) => {
+		for (let i = 0; i < this._pendingReplaceTracks.length; i++) {
+			if (this.pc.iceConnectionState === 'new') {
+				// Do not replace the tracks when the connection has not started
+				// yet, as Firefox can get "stuck" and not replace the tracks
+				// even if tried later again once connected.
+
+				resolve(false)
+
+				return
+			}
+
+			const pending = this._pendingReplaceTracks[i]
+
+			try {
+				this._replaceTrack(pending.newTrack, pending.oldTrack, pending.stream).then(() => {
+					this._pendingReplaceTracks.splice(i, 1)
```

Comment:
this feels dangerous.

we're tying the callback to a fixed index but maybe the track list will have shifted then, for example if the callbacks are called out of order or delayed ? in case of shift, you'd be deleting the wrong index.

also it's a bit difficult to grasp how this connects to the main promise from this function

Reply:
This should be fixed with the reintroduction of the `await` call.